src/formats/SMI.cpp

```diff
@@ -441,13 +394,88 @@ static void find_rings(
+        // Not guaranteed to fast, nor efficient, but just to work
+        // Also not guaranteed to find the SSSR, but it will find all rings
+        // needed for a given structure to be writen as SMILES
+        // adj_list The adjacentcy list graph
+        // hit_atoms Atoms already encountered
+        // ring_bonds Bonds already known to ring forming bonds
+        // ring_atoms Map from an atom to the number of rings it forms
+
+        // The atoms to process consist of :
+        // current_atom Atom to be checked in this iteration
+        // previous_atom Atom which was bound to the current atom
+
+        std::stack<std::pair<size_t, size_t>> atoms_to_process;
+        atoms_to_process.push({ current_atom, current_atom });
+
+        while (atoms_to_process.size()) {
+
+            size_t previous_atom;
+
+            std::tie(current_atom, previous_atom) = atoms_to_process.top();
+            atoms_to_process.pop();
+
+            auto& current_atom_bonds = adj_list[current_atom];
+
+            // We go through the neighbors backwards because we are using a
+            // stack instead and want to through them in the forward direction
+            // (think of it like the negative of a negative is a positive)
+            // We cannot use a queue as we want to process all neighbors of an
+            // atom first, then move on to the next atom group (allowing for a
+            // depth first like search).
+            for (auto neighbor_iter = current_atom_bonds.rbegin();
+                neighbor_iter != current_atom_bonds.rend();
+                ++neighbor_iter) {
+
+                auto neighbor = *neighbor_iter;
+
+                // prevent back tracking
+                if (neighbor == previous_atom) {
+                    continue;
+                }
+
+                // We've seen this neighbor before! Ring found
+                // but only if we have NOT processed the current atom as this
+                // prevents rings from being created from mutliple directions.
+                if (hit_atoms[neighbor] && !hit_atoms[current_atom]) {
+
+                    // Don't process the same ring connection twice
+                    if (ring_bonds.count(Bond(neighbor, current_atom))) {
+                        continue;
+                    }
+                    ring_bonds.insert(Bond(neighbor, current_atom));
+
+                    auto neigh_ring_iter = ring_atoms.find(neighbor);
+                    if (neigh_ring_iter == ring_atoms.end()) {
+                        neigh_ring_iter = ring_atoms.insert({ neighbor, 0 }).first;
+                    }
+                    neigh_ring_iter->second++;
+
+                    continue; // No need to see this atom again
+                }
+
+                // Continue the search for new atoms
+                if (!hit_atoms[neighbor]) {
+                    //find_rings_helper(adj_list, hit_atoms, ring_bonds, ring_atoms, neighbor, current_atom);
```

Comment:
left-over code?

Reply:
Yes, it is. When rewriting the function I wanted to keep track of where the recursion was.